internal/mux/buffer_test.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 The Oto Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mux_test
+
+import (
+	"io"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/hajimehoshi/oto/internal/mux"
+)
+
+const concurrency = 1000
+
+func TestBufferConcurrentWrites(t *testing.T) {
+	b := mux.NewConcurrentBuffer()
+
+	ch := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		go func() {
+			<-ch
+			b.Write([]byte{1})
+			wg.Done()
+		}()
+	}
+	close(ch)
+	wg.Wait()
+
+	if l := b.Len(); l != concurrency {
+		t.Errorf("Expected Len to be %v, but it was %v", concurrency, l)
+	}
+
+	out := make([]byte, concurrency)
+	n, err := b.Read(out)
+	if err != nil {
+		t.Errorf("Expected no error from Read, but it returned: %v", err)
+	}
+	if n != concurrency {
+		t.Errorf("Expected Read to return %v bytes, but it returned %v", concurrency, n)
```

Comment:
`read bytes: got: %v, want: %v`

Reply:
BTW, `io.Read` doesn't necessarily return `len(out)`, then this test depends on implementation detail. Use `io.ReadAll`.